src/common/snippets/src/op/buffer.cpp

```diff
@@ -13,89 +13,83 @@ namespace ov {
-    return true;
-}
-
-bool Buffer::is_defined() const {
-    return !utils::is_dynamic_value(m_allocation_size);
-}
-
-size_t Buffer::get_byte_size() const {
-    if (is_defined())
-        return m_allocation_size * get_element_type().size();
-    return utils::get_dynamic_value<size_t>();
-}
+Buffer::Buffer(const ov::Output<ov::Node>& arg) : Buffer(ov::OutputVector{arg}) {}
 
-IntermediateMemoryBuffer::IntermediateMemoryBuffer(const OutputVector& arguments, size_t allocation_size, size_t reg_group, size_t cluster_id)
-    : Buffer(arguments, allocation_size, reg_group, cluster_id) {
+Buffer::Buffer(const OutputVector& arguments) : Op(arguments), m_type(Type::IntermediateMemory) {
     constructor_validate_and_infer_types();
 }
 
-IntermediateMemoryBuffer::IntermediateMemoryBuffer(const ov::Output<ov::Node>& arg, size_t allocation_size, size_t reg_group, size_t cluster_id)
-    : IntermediateMemoryBuffer(OutputVector{arg}, allocation_size, reg_group, cluster_id) {}
-
-void IntermediateMemoryBuffer::validate_and_infer_types() {
-    INTERNAL_OP_SCOPE(Buffer_validate_and_infer_types);
-    ov::PartialShape output_shape;
-    set_output_type(0, get_input_element_type(0), get_input_partial_shape(0));
-}
-
-std::shared_ptr<Node> IntermediateMemoryBuffer::clone_with_new_inputs(const OutputVector& new_args) const {
-    INTERNAL_OP_SCOPE(Buffer_clone_with_new_inputs);
-    auto new_buffer = std::make_shared<IntermediateMemoryBuffer>(new_args, m_allocation_size, m_reg_group, m_cluster_id);
-    new_buffer->set_offset(m_offset);
-    return new_buffer;
+Buffer::Buffer(const ov::Shape& shape, ov::element::Type element_type) : Op(), m_type(Type::NewMemory), m_output_shape(shape), m_element_type(element_type) {
+    constructor_validate_and_infer_types();
 }
 
-NewMemoryBuffer::NewMemoryBuffer(const ov::Shape& shape, size_t reg_group, size_t cluster_id, ov::element::Type element_type)
-    : Buffer({}, ov::shape_size(shape), reg_group, cluster_id), m_output_shape(shape), m_element_type(element_type) {
-    constructor_validate_and_infer_types();
+bool Buffer::visit_attributes(AttributeVisitor& visitor) {
+    INTERNAL_OP_SCOPE(Buffer_visit_attributes);
+    auto shape = utils::pshape_to_vdims(get_output_partial_shape(0));
+    auto etype = get_output_element_type(0);
+    visitor.on_attribute("shape", shape);
+    visitor.on_attribute("element_type", etype);
+    return true;
 }
 
-void NewMemoryBuffer::validate_and_infer_types() {
+void Buffer::validate_and_infer_types() {
     INTERNAL_OP_SCOPE(Buffer_validate_and_infer_types);
-    OPENVINO_ASSERT(get_input_size() == 0, "Buffer with new allocated memory mustn't have arguments!");
-    set_output_type(0, m_element_type, m_output_shape);
+    if (m_type == Type::NewMemory) {
+        OPENVINO_ASSERT(get_input_size() == 0, "NewMemory Buffer mustn't have inputs");
+        set_output_type(0, m_element_type, m_output_shape);
+    } else if (m_type == Type::IntermediateMemory) {
```

Comment:
Which benefits do we get from combining NewMemory and IntermediateMemory buffers into one class? As a drawback, we now have to insert `if`s in several places in class logic. When we separated buffers into 2 classes, one of the main motivation was avoiding such `if`s.
This unification confuses me, and I believe that separate classes are better. Can we discuss this?

Reply:
The idea was:
1. One common `BufferExpression` - one common `op::Buffer`.
2. The user shouldn't think which `op::Buffer` he need to use - just call the corresponding ctor.

I agree that now we have many `if` in the class but it's hidden from the user (he don't know about them).
And this `type` was implemented just to easier implement methods (in general we can use `if (get_input_size() == 0)` but I decided to implement private types instead of that)

I agree that we should discuss it one more time. Let's do it offline!

Reply:
Discussed offline:
Decided to implement new class field `m_impl` which responses on different behaviors and encapsulate this logic. There were two ideas:
1. Create two classes `NewMemory` and `IntermediateMemory` derived from `ov::Op` and store the object as smart pointer in attributes of `op::Buffer`. I investigated that this idea is not safe because if the user call `set_argument` for the buffer, we must call this method for `m_impl` also. However, this method is not virtual and cannot be correctly overrided - we will get incorrect behavior in this case.
2. Create two classes `NewMemory` and `IntermediateMemory` derived from `ov::Buffer` and store the object as smart pointer in attributes of `op::Buffer`. There is the same problem as described previously.

So I implemented the one abstract base class `BaseImpl` with needed abstract methods and 2 derived the corresponding classes. Now it looks better (at least for me).

Could you take a look please one more time?

Reply:
From my perspective, the new implementation much better. Thank you!